packages/material-ui-lab/src/MasonryItem/MasonryItem.js

```diff
@@ -0,0 +1,154 @@
+  const { classes } = styleProps;
+
+  const slots = {
+    root: ['root'],
+  };
+
+  return composeClasses(slots, getMasonryItemUtilityClass, classes);
+};
+
+export const style = ({ styleProps, theme }) => {
+  let styles = {
+    width: '100%',
+    [`& > *`]: {
+      // all contents should have a width of 100%
+      objectFit: 'cover',
+      width: '100%',
+    },
+    visibility: styleProps.contentHeight ? 'visible' : 'hidden',
+    ...((typeof styleProps.spacing === 'string' || typeof styleProps.spacing === 'number') && {
+      gridRowEnd: `span ${
+        styleProps.contentHeight + Number(theme.spacing(styleProps.spacing).replace('px', ''))
+      }`,
+      paddingBottom: Number(theme.spacing(styleProps.spacing).replace('px', '')) - 1,
+    }),
+  };
+
+  if (typeof styleProps.spacing === 'object' || Array.isArray(styleProps.spacing)) {
+    const base = Object.keys(theme.breakpoints.values).reduce((acc, breakpoint) => {
+      if (styleProps.spacing[breakpoint] != null) {
+        acc[breakpoint] = true;
+      }
+      return acc;
+    }, {});
+    const spacingValues = resolveBreakpointValues({ values: styleProps.spacing, base });
+    const transformer = createUnarySpacing(theme);
+    const styleFromPropValue = (propValue) => {
+      const gap = Number(getValue(transformer, propValue).replace('px', ''));
+      const rowSpan = styleProps.contentHeight ? Math.ceil(styleProps.contentHeight + gap) : 0;
+      return {
+        gridRowEnd: `span ${rowSpan}`,
+        paddingBottom: gap - 1,
+      };
+    };
+    styles = deepmerge(styles, handleBreakpoints({ theme }, spacingValues, styleFromPropValue));
+  }
+  return styles;
+};
+
+const MasonryItemRoot = styled('div', {
+  name: 'MuiMasonryItem',
+  slot: 'Root',
+  overridesResolver: (props, styles) => {
+    return [styles.root];
+  },
+})(style);
+
+const MasonryItem = React.forwardRef(function MasonryItem(inProps, ref) {
+  const props = useThemeProps({
+    props: inProps,
+    name: 'MuiMasonryItem',
+  });
+
+  const masonryItemRef = React.useRef(null);
+  const handleRef = useForkRef(ref, masonryItemRef);
+  const {
+    spacing = 1,
+    documentReady = false,
+  } = React.useContext(MasonryContext);
+  const { children, className, component = 'div', ...other } = props;
+  const [styleProps, setStyleProps] = React.useState({
+    ...props,
+    spacing,
+  });
+
+  const classes = useUtilityClasses(styleProps);
+
+  const computeHeight = () => {
+    const child = masonryItemRef.current.firstChild;
+    setStyleProps({
+      ...styleProps,
+      contentHeight: child?.getBoundingClientRect().height,
+    });
+  };
+
+  React.useEffect(() => {
+    if (documentReady) {
+      computeHeight();
+    }
+  }, [documentReady]); // eslint-disable-line
+
+  React.useEffect(() => {
+    window.addEventListener('resize', computeHeight);;
+    return () => {
+      window.removeEventListener('resize', computeHeight);
+    };
+  });
+  return (
```

Comment:
We can use a `useResizeObserver` instead, see https://github.com/mui-org/material-ui/pull/23954#issuecomment-747411562